src/modeling_library/unfold/backprop.jl

```diff
@@ -26,17 +66,21 @@ function accumulate_param_gradients!(trace::VectorTrace{UnfoldType,T,U}, retval_
         end
     end
 
-    for key=1:len
+    local kernel_arg_grads::Tuple = (nothing, nothing) # initialize for (len index, state)
+    for key = len : -1 : 1 # walks backward over chain
         subtrace = trace.subtraces[key]
         kernel_retval_grad = (retval_grad == nothing) ? nothing : retval_grad[key]
-        kernel_arg_grad::Tuple = accumulate_param_gradients!(subtrace, kernel_retval_grad)
-        @assert kernel_arg_grad[1] == nothing
-        @assert kernel_arg_grad[2] == nothing
-        for (i, (grad, has_grad)) in enumerate(zip(kernel_arg_grad[3:end], params_has_grad))
+        if state_has_grad
+            kernel_retval_grad = fold_sum(kernel_retval_grad, kernel_arg_grads[2])
+        end
+        kernel_arg_grads = accumulate_param_gradients!(subtrace, kernel_retval_grad)
+        @assert kernel_arg_grads[1] == nothing
+        state_has_grad || @assert kernel_arg_grads[2] == nothing
+        for (i, (grad, has_grad)) in enumerate(zip(kernel_arg_grads[3:end], params_has_grad))
             if has_grad
                 params_grad[i][key] = grad
             end
         end
     end
-    (nothing, nothing, params_grad...)
+    (nothing, kernel_arg_grads[2], params_grad...)
```

Comment:
I think there was previously a bug here involving `params_grad` (they need to be summed across 'keys' and they are not), and this bug remains. Also, the corresponding test code was and remains commented out: 

        #@test isapprox(input_grads[3], expected_xs_grad) # alpha
        #@test isapprox(input_grads[4], expected_ys_grad) # beta
        #@test isapprox(input_grads[4], expected_ys_grad) # beta